### PR TITLE
[compiler] Rewrite StreamTake and StreamDrop to TakeWhile and DropWhile

### PR DESCRIPTION
**Pull request by tpoterba**
Also add specialized code generators for range/zipWithIndex

### CODE REVIEW

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
This is great! A couple minor things, and some semantics questions.

File: hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala [hunk 1]
```diff
@@ -1743,6 +1897,56 @@ object EmitStream {
           }
         }
 
+      case StreamZip(
+      IndexedSeq(s1, StreamRange(I32(0), I32(Int.MaxValue), I32(1), _, _)),
+      IndexedSeq(prevName, idxName),
+      body,
+      ArrayZipBehavior.TakeMinLength, _) =>
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala:1904]: Could you elaborate on what you think the bad interaction would be?

**Comment by tpoterba** [comment]
I backed out the iota optimzations to make this PR less controversial -- it now only introduces and tests StreamTakeWhile and StreamDropWhile. The other functionality will be a stacked PR.

File: hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala [hunk 2]
```diff
@@ -657,6 +713,104 @@ object EmitStream {
...
+
+            SStreamCode(producer)
+          }
+
+      case StreamDropWhile(a, elt, condIR) =>
+        produce(a, cb)
+          .map(cb) { case (childStream: SStreamCode) =>
+            val childProducer = childStream.producer
+            val eltSettable = mb.newEmitField("stream_take_while_elt", childProducer.element.emitType)
+            val doneComparisons = mb.genFieldThisRef[Boolean]("stream_take_while_donecomparisons")
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala:707]: Still have a couple typos here

File: hail/src/main/scala/is/hail/expr/ir/IR.scala [hunk 3]
```diff
@@ -229,6 +229,11 @@ final case class ArrayRef(a: IR, i: IR, errorID: Int) extends IR
 final case class ArraySlice(a: IR, start: IR, stop: Option[IR], step:IR = I32(1), errorID: Int = ErrorIDs.NO_ERROR) extends IR
 final case class ArrayLen(a: IR) extends IR
 final case class ArrayZeros(length: IR) extends IR
+
+object StreamRange {
+  def iota(requiresMemoryManagement: Boolean = false): StreamRange = StreamRange(0, Int.MaxValue, 1, requiresMemoryManagement)
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/IR.scala:234]: Missed this and `zipWithIndex`, but it's fine to just fix in the follow-up PR.

**Review by patrick-schultz — DISMISSED** [review]
Looks good, just one copy-paste typo.

File: hail/src/main/scala/is/hail/expr/ir/IR.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/IR.scala:234]: oops, thanks!

**Comment by tpoterba** [comment]
thanks, addressed!

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**